File: src/perl/pangenome_make_table.pl
#!/usr/bin/perl

=head1  NAME

pangenome_make_table.pl - Creates counts of new and core genes per pangenome size

=head1 SYNOPSIS

USAGE: pangenome_make_table.pl
        --input_list=/path/to/somefile.dat.list
        --output=/path/to/output.list
      [ --log=/path/to/some/log ]

=head1 OPTIONS

B<--blast_stored_file,-b>
    List of files containing serialized array of BLAST data.

B<--output_path,-o>
    Path to which output files will be written.

B<--output_file,-o>
    Full path and name of pangenome data output file.

B<--multiplicity,-m>
    The multiplicity value to be used in data sampling.

B<--comparisons,-c>
    The number of comparisons to sample.

B<--num_threads,-t>
    Optional.  Enable multithreading, using the specified number of threads.

B<--log,-d>
    optional. Will create a log file with summaries of all actions performed.

B<--debug>
    optional. Will display verbose status messages to STDERR if logging is disabled.

B<--help,-h>
    This help message/documentation.

=head1   DESCRIPTION

    The pangenome analysis script creates an array of BLAST results data which is then
    processed to create pangenome data.

=head1 INPUT

    The input should be a list of files containing serialized BLAST results array data.

=head1 OUTPUT

    There is no output unless you use the --log option.

=head1 CONTACT

	Shaun Adkins
	sadkins@som.umaryland.edu

=cut

use Pod::Usage;
use File::Basename;
use XML::Twig;
use Math::Combinatorics;
use Getopt::Long qw(:config no_ignore_case no_auto_abbrev);
use Storable qw(nstore retrieve);
use Data::Random qw(:all);    ## temp
use Benchmark;
use bignum;
#use DBM::Deep;

use threads;    # SAdkins - 7/22/16 - Discouraged by Perl but let's try it!
use Thread::Queue;
use Thread::Semaphore;

use warnings;
use strict;

my $dups    = ();
my %options = ();
my @threads;

### Sampling analysis hashes
my $comp_counter = ();
my $dup_counts = ();
my $genes_by_category = ();

my $results = GetOptions(
    \%options,

    #                           'filter_list|f=',
    'db_list|dl:s',
    'blast_stored_file|b=s',
    'output_path|o=s',
    'output_file|f=s',
    'write_lists:i',
    'multiplicity|m:s',
    'comparisons|c:s',
    'num_threads|t:s',
    'debug|d=s',
    'log|l=s',
    'help|h',
) || pod2usage();

my $comparisons;
my $multiplicity;
my $db_filter = undef;

my $max_threads = $options{'num_threads'} ? $options{'num_threads'} : 1;
print STDERR "Max number of threads used: $max_threads\n";
#my @threads = init_threads($max_threads);

if ( $options{'db_list'} ) {
    &read_db_list();
}

unless ( $options{'comparisons'} || $options{'multiplicity'} ) {

    #    die "must provide flag --comparisons or --multiplicity";
    print STDERR "Will run analysis without sampling\n";
}

my $output_path = $options{'output_path'};
$output_path =~ s/\/$//;
if ( $output_path eq '' ) {
    $output_path = '.';
}

my $output_file;
if ( $options{'output_file'} ) {
    $output_file = $options{'output_file'};
} else {
    $output_file = 'pangenome.table.txt';
}

if ( !-e $options{'blast_stored_file'} ) {
    die "must specify an input list which exists";
}

##open (IN, $options{'input_list'}) || die "couldn't open input list";

print STDERR "Reading stored data...";
if ( !-e $options{'blast_stored_file'} ) {
    die "no stored blast data file found in output dir";
}

my $temp_ref = retrieve( $options{'blast_stored_file'} );
## pull the blast results data array out of the stored array
my $results_ref = shift( @{$temp_ref} );

## pull the dups hash out of the stored array
my $dups_ref = shift( @{$temp_ref} );
$dups = $dups_ref;

print STDERR "done.\n";

my %dbs;        #Keeps track of organism name entries
my @genomes;    #Also keeps track of org entries
my $genes = (); # keeps track of query_db -> query_prot -> subject_db

#my $genes = DBM::Deep->new( ".pangenome.temp.db" );

print STDERR "Processing results...";
# Create profile hash structure from the binary input structure
foreach my $qgenome (keys %$results_ref) {
    next if ($db_filter && !$db_filter->{$qgenome});    # skip genome if not in optional filter list
    $genes->{$qgenome} = ();
    $dbs{$qgenome} = 1;
    foreach my $qgene (keys %{$results_ref->{$qgenome}}) {
        $genes->{$qgenome}->{$qgene} = {};
        foreach my $sgenome (@{$results_ref->{$qgenome}->{$qgene}}) {
            next if ($db_filter && !$db_filter->{$sgenome});
            $genes->{$qgenome}->{$qgene}->{$sgenome} = 1;
        }
    }
}
@genomes = keys(%dbs);

open( RESULT, ">" . $output_path . "/" . $output_file )
  || die "couldn't open $output_file for writing";

my %genome_index = ();
my $i_counter    = 1;
foreach (@genomes) {
    $genome_index{$_} = $i_counter;

    #print RESULT "## $i_counter\t$_\n";
    $i_counter++;
}

# output the gene counts for single genomes
foreach (@genomes) {
    my $genome_dup_count = 0;
    my $gene_count       = scalar keys( %{ $genes->{$_} } );
    foreach my $dup_set ( keys( %{ $dups->{$_} } ) ) {
        my @dup_genes = split( " ", $dup_set );
        $genome_dup_count += scalar(@dup_genes);
    }
    ## our output table will take the form:
    ## genomecount\tcore\tshared\tnew\tcore dup count\tshared dup\tnew dup
    print RESULT "1\t0\t0\t$gene_count\t0\t0\t$genome_dup_count\t"
      . $genome_index{$_}
      . "\t$_\n";
}
print STDERR "done.\n";
if ( $options{'comparisons'} || $options{'multiplicity'} ) {
    &do_analysis_with_sampling();
} else {
    # Will take a lot longer to finish due to doing all possible combinations
    &do_analysis_no_sampling();
}

exit(0);

sub do_analysis_with_sampling {
    if ( $options{'comparisons'} ) {
        my ( $est_multiplicity, $t, $t2 ) =
          estimate_multiplicity( scalar(@genomes), $options{'comparisons'} );
        print STDERR "# of comparisons: $options{comparisons}\nest. multiplicity: $est_multiplicity\n";
        $options{'multiplicity'} = $est_multiplicity;
    } else {
        my ( $est_comparisons, $t ) =
          estimate_comparisons( scalar(@genomes), $options{'multiplicity'} );
        print STDERR "multiplicity: $options{multiplicity}\nest. # comparisons: $est_comparisons\n";
    }

    $comp_counter = ();

    my $max = scalar @genomes;

    print STDERR "Total of $max genomes in results set\n";
    print STDERR "Running analysis...\n";

    my %seen;

    my @i_genomes = ( 0 .. ( $max - 1 ) );

    # Iterate using combinations of reference genomes of size $i
    for ( my $i = 1; $i < $max; $i++ ) {

        # start timer
        my $start = new Benchmark;

        print STDERR "Running analysis with $i reference genomes...";

        ## new
        my $true_max =
          int( factorial( $max - 1 ) /
              ( factorial( ( $i + 1 ) - 1 ) * factorial( $max - ( $i + 1 ) ) )
          );
#        print STDERR "Maximum number of computations for $i on $max genomes: $true_max\n";

        # Iterate through each available genome
        for ( my $j = 0; $j < $max; $j++ ) {
            #Get comparison genome name and index number
            my $comp_genome = $genomes[$j];

            my @ref_genomes = @i_genomes;
            splice @ref_genomes, $j, 1;

            # Allocated concurrent thread limit
#           my $sem = Thread::Semaphore->new($max_threads);

            my $point_count = 0;

#           @threads = map { threads->create( sub {
#               $sem->down();   # Note resource is being used
                # Iterate until either the number of multiplicity rounds is reached, or all possible combinations of ref genomes is seen
                while ( $point_count < min($options{'multiplicity'}, $true_max) ) {
                    # Choose a random set of N reference genomes
                    my @reference_set =
                      rand_set( set => \@ref_genomes, size => $i, shuffle => 0 );

					# Create a vector of genome ids among the reference set
                    my @seen_vec = ();
                    for ( my $ii = 0; $ii < scalar(@reference_set); $ii++ ) {
                        push @seen_vec, $reference_set[$ii];
                    }
					# Sort the vector, numerically, and create a string
					my @sorted_vec = sort {$a <=> $b} @seen_vec;
					my $sort_vec_string = join(':', @sorted_vec);

                    unless ( defined $seen{$j}{$sort_vec_string} ) {
                        $dup_counts = ();
                        # Counter to track number of iterations of comparison genome in use per size-$i ref genomes
                        $comp_counter->{$comp_genome}->{$i}++;
                        $genes_by_category = ();

                        # Iterate through the genes in the query genome
                        GENE:
                        foreach my $gene ( keys( %{ $genes->{$comp_genome} } ) ) {
                            # Iterate through our N reference genome indexes
                            foreach my $i_ref_genome (@reference_set) {
                                # grabbing the subject db name
                                my $ref_genome = $genomes[$i_ref_genome];
                                my $hit = $genes->{$comp_genome}->{$gene}->{$ref_genome};
                                categorize_shared_gene($hit, $comp_genome, $gene);
                            }

                            my $count = (defined $genes_by_category->{$comp_genome}->{'shared'}->{$gene}) ? $genes_by_category->{$comp_genome}->{'shared'}->{$gene} : 0;
                            categorize_core_gene($comp_genome, $gene) if $count == scalar @reference_set;
                            categorize_new_gene($comp_genome, $gene) if $count == 0;
                          } #/GENE

                          count_dups($comp_genome);
                          write_results($comp_genome, \@reference_set, $i);

                          ## Ensure we don't use this set of genomes again
                          $seen{$j}{$sort_vec_string} = 1;
						  # Only increment point count if we had a unique reference genome set
						  $point_count++;
                    } ## end of unless
                } ## end of while 'point_count'
#                $sem->up(); # Free up resource
#            }); ## end of thread->create sub
#            } ## end of threads
#            $_->join() for @threads;
        }
        # end timer
        my $end = new Benchmark;

        # calculate difference
        my $diff = timediff( $end, $start );

        print STDERR "runtime: " . timestr( $diff, 'noc' ) . "\n";
    }
}

sub do_analysis_no_sampling {
    $comp_counter = ();
    my $max          = scalar @genomes;

    # Get combinations of reference genomes of size $i
    for ( my $i = 1; $i < $max; $i++ ) {
        # start timer
        my $start = new Benchmark;

        my $combinat = Math::Combinatorics->new(
            count => $i,
            data  => [@genomes],
        );

        # For each combination of reference genomes...
        while ( my @reference = $combinat->next_combination ) {
            my $ref_string = "(" . join( ",", @reference ) . ")\n";

            # Allocated concurrent thread limit
            my $sem = Thread::Semaphore->new($max_threads);            # Grab the genomes not in our reference set.
            my @comparison_set = @{ array_diff( \@genomes, \@reference ) };

            @threads = map { threads->create( sub {
                $sem->down();   # Note resource is being used
                # Each comparison genome will be compared to the reference genomes
                foreach my $comp_genome (@comparison_set) {
                    $dup_counts = ();
                    $comp_counter->{$comp_genome}->{$i}++;
                    $genes_by_category = ();

                    # Process each gene from the comparison genome
                    GENE:
                    foreach my $gene ( keys( %{ $genes->{$comp_genome} } ) ) {
                        # Keep track of how many ref genomes had a hit for this particular gene
                        foreach my $ref_genome (@reference) {
                            my $hit = $genes->{$comp_genome}->{$gene}->{$ref_genome};
                            categorize_shared_gene($hit, $comp_genome, $gene);
                        }
                        my $count = (defined $genes_by_category->{$comp_genome}->{'shared'}->{$gene}) ? $genes_by_category->{$comp_genome}->{'shared'}->{$gene} : 0;
                        categorize_core_gene($comp_genome, $gene) if $count == scalar @reference;
                        categorize_new_gene($comp_genome, $gene) if $count == 0;

                    } # /GENE

                   count_dups($comp_genome);
                   write_results($comp_genome, \@reference, $i);
                }
                $sem->up(); # Free up resource
            });
            }
            $_->join() for @threads;
        }
        # end timer
        my $end = new Benchmark;

        # calculate difference
        my $diff = timediff( $end, $start );

        print "runtime: " . timestr( $diff, 'noc' ) . "\n";
    }
}

# Return the minimum of the two values
sub min {
    my ($a, $b) = @_;
    return $a < $b ? $a : $b;
}

# Return an array of elements present in array A but not in array B
sub array_diff {
    my ( $superset_ref, $subset_ref ) = @_;
    my %index;
    my @diff_array;
    foreach ( @{$superset_ref} ) {
        $index{$_} = 1;
    }
    foreach ( @{$subset_ref} ) {
        delete $index{$_};
    }
    @diff_array = keys(%index);
    return \@diff_array;
}

no warnings;    # Guards against 'deep recursion' warnings
sub factorial {
    my $i = shift @_;
    if ( $i <= 1 ) {
        return 1;
    } else {
        return ( $i * factorial( $i - 1 ) );
    }
}
use warnings;

sub estimate_multiplicity {
    my ( $ngenomes, $req_comparisons ) = @_;
    my $i           = 0;
    my $lower_mult  = 0;
    my $lower_comp  = 0;
    my $lower_theor = 0;
    my $upper_mult  = 0;
    my $upper_comp  = 0;
    my $upper_theor = 0;
    my $ldiff       = 0;
    my $udiff       = 0;
  LOOP: for ( $i = 5; $i <= 5000; $i++ ) {
        my ( $a, $b ) = estimate_comparisons( $ngenomes, $i );
        if ( $a < $req_comparisons ) {
            $lower_mult  = $i;
            $lower_comp  = $a;
            $lower_theor = $b;
        } else {
            $upper_mult  = $i;
            $upper_comp  = $a;
            $upper_theor = $b;
            last LOOP;
        }
    }
    if ( $upper_mult == 5 ) {
        my ( $a, $b ) = estimate_comparisons( $ngenomes, 5 );
        return ( 5, $a, $b );
    } elsif ( $lower_mult == 5000 ) {
        return ( 5000, $lower_comp, $lower_theor );
    } else {
        $ldiff = $req_comparisons - $lower_comp;
        $udiff = $upper_comp - $req_comparisons;
        if ( $ldiff <= $udiff ) {
            return ( $lower_mult, $lower_comp, $lower_theor );
        } else {
            return ( $upper_mult, $upper_comp, $upper_theor );
        }
    }
}

sub estimate_comparisons {
    my ( $ngenomes, $multiplex ) = @_;

    my $i                 = 0;
    my $tot_comparisons   = 0;
    my $theor_comparisons = 0;
    my $theor             = 0;
    my $real              = 0;
    for ( $i = 2; $i <= $ngenomes; $i++ ) {
        my $theor = factorial($ngenomes) /
          ( factorial( $i - 1 ) * factorial( $ngenomes - $i ) );
        my $real = $multiplex * $ngenomes;
        $theor_comparisons += $theor;
        $tot_comparisons += min($theor, $real);
    }

    return ( $tot_comparisons, $theor_comparisons );
}

sub read_db_list {
    open IN, "<" . $options{'db_list'}
      || die "couldn't open '$options{db_list}' for reading: $!";
    while (<IN>) {
        chomp;
        $db_filter->{$_} = 1;
    }
    close IN;

}

# Subroutine to force a specific size for the max number of threads
sub init_threads {
    # An array to place our threads in
    my $num_of_threads = shift;
	my @initThreads;
	for(my $i = 1;$i<=$num_of_threads;$i++){
		push(@initThreads,$i);
	}
	return @initThreads;
}

# Determine if the given gene is also present in another genome up to this point.
sub categorize_shared_gene {
    my ($hit, $comp_genome, $gene) = @_;
    ## increment shared count for this gene if we have a hit
    $genes_by_category->{$comp_genome}->{'shared'}->{$gene}++ if $hit;
}

# Determine if the given gene is present in all other genomes up to this point.
sub categorize_core_gene {
    my ($comp_genome, $gene) = @_;
    $genes_by_category->{$comp_genome}->{'core'}->{$gene} = 1;
}

# Determine if the given gene is present in no other genomes up to this point.
sub categorize_new_gene {
    my ($comp_genome, $gene) = @_;
    $genes_by_category->{$comp_genome}->{'new'}->{$gene} = 1;
}

# Count numbers of core, shared, and new genes amongst duplicate hits
sub count_dups {
    my $comp_genome = shift;
    ## process lists to see how many duplicated genes are in each category
    foreach my $cat ( ( 'core', 'shared', 'new' ) ) {
        $dup_counts->{$comp_genome}->{$cat} = 0;
        ## we'll look at each set of duplicates one at a time
        foreach
          my $dup_set ( keys( %{ $dups->{$comp_genome} } ) )
        {
            my @dup_genes   = split( " ", $dup_set );
            my $dup_count   = scalar(@dup_genes);
            my $dup_counter = 0;

            ## for each gene of a duplicate set
            foreach my $dup (@dup_genes) {
                ## check if it's in the category
                if ( $genes_by_category->{$comp_genome}->{$cat}
                    ->{$dup} )
                {
                    $dup_counter++;
                }
            }
            ## check if all of the dups of a set weren't found in the same category
            if (   $dup_counter > 0
                && $dup_count != $dup_counter )
            {
                print STDERR
                  "Only $dup_counter of the following dup set found:\n$dup_set\n***This could be a problem: Comp_genome $comp_genome\n";
                ## and if they are, then add the dup overcount to the total
            } elsif ( $dup_counter == $dup_count ) {
                $dup_counts->{$comp_genome}->{$cat} +=
                  $dup_count;
            }
        }
    }
}

# Write the results to file
sub write_results {
    my $comp_genome = shift;
    my $ref = shift;
    my $size = shift;

    my $core_count =
      scalar(
        keys( %{ $genes_by_category->{$comp_genome}->{'core'} } ) );
    my $shared_count =
      scalar(
        keys( %{ $genes_by_category->{$comp_genome}->{'shared'} } ) );
    my $new_count =
      scalar(
        keys( %{ $genes_by_category->{$comp_genome}->{'new'} } ) );

    my $core_dup_count   = $dup_counts->{$comp_genome}->{'core'};
    my $shared_dup_count = $dup_counts->{$comp_genome}->{'shared'};
    my $new_dup_count    = $dup_counts->{$comp_genome}->{'new'};

    # This equals the number of references plus our one comparison genome
    my $rgcount = ( scalar @$ref ) + 1;

    print RESULT
      "$rgcount\t$core_count\t$shared_count\t$new_count\t$core_dup_count\t$shared_dup_count\t$new_dup_count\t"
      . $genome_index{$comp_genome}
      . "\t$comp_genome\n";

# Would love to make this its own subroutine eventually but don't feel like dealing with all the extra variables
    if ( $options{'write_lists'} ) {
        write_list("core", $size, $rgcount, $ref, $comp_genome);
        write_list("new", $size, $rgcount, $ref, $comp_genome);
        write_list("shared", $size, $rgcount, $ref, $comp_genome);
    }
}

sub write_list {
    my ($group, $size, $rgcount, $ref, $comp_genome) = @_;
    open( OUT,
            ">"
          . $output_path . "/"
          . $comp_genome . "_${group}_"
          . $rgcount . "_"
          . $comp_counter->{$comp_genome}->{$size} );
    print OUT "#" . $comp_genome . "\n";
    print OUT "#" . join( " ", @$ref ) . "\n";
    foreach my $g (keys(%{ $genes_by_category->{$comp_genome}->{$group} })){
        print OUT $g . "\n";
    }
    close OUT;
}

File: src/julia/pangenome_make_table.jl
#!/usr/bin/julia

# pangenome_make_table.jl - Reads a pangenome profile and generates core, new, and shared gene hits for the pangenome
# By: Shaun Adkins (sadkins@som.umaryland.edu)

using ArgParse

#############
# Type defs #
#############

type HitsProfile
    profile::Dict
    genome_names::Vector{String}
    num_genomes::Integer
end

HitsProfile() = HitsProfile(Dict(), Vector{String}(), 0)

###############
# HitsProfile #
###############

function read_in_profile(hp::HitsProfile, profile)
    # Read in tab-delimited file, and grab genome names
    (data, headers) = readdlm(profile, '\t', header=true, comments=false)
    hp.num_genomes = length(headers) - 2
    hp.genome_names = headers[3:end]
    hp.profile = parse_data(hp.genome_names, data)
end

function parse_data(headers::Vector{String}, data)
    # Iterate through each line of data
    profile = Dict{String, Any}()
    for i = 1:size(data,1)
        genome = data[i,1]
        gene = data[i,2]
        # Hits to each subject genome
        hits = data[i,3:end]
        g = Dict{String, Bool}()
        for j = 1:length(hits)
            if hits[j] == 1
                g[headers[j]] = true
            end
        end

        # Create typed dictionary for genome if one does not exist
        try
			profile[genome][gene] = g
		catch KeyError e
			profile[genome] = Dict{String, Any}(gene => g)
		end
    end
    return profile
end

########
# Main #
########

function determine_sampling_rate( hp::HitsProfile, args)
    if args["comparisons"] > 0
        return estimate_multiplicity(hp.num_genomes, args["comparisons"])
    else
        return args["multiplicity"]
    end
end

function estimate_multiplicity( genomes::Integer, comparisons::Integer)
    # Determine multiplex sampling rate based on specified comparisons
    lower_mult = 0
    lower_comp = 0
    lower_theor = 0
    upper_mult = 0
    upper_comp = 0
    upper_theor = 0
    ldiff = 0
    udiff = 0

	# Bound limits for estimating multiplicity rate
	lower_bounds = 5
	upper_bounds = 5000

    # Establish multiplex bounds
    for i = lower_bounds:upper_bounds
        (est_comps, theor_comps) = estimate_comparisons(genomes, i)
        if est_comps < comparisons
            lower_mult = i
            lower_comp = est_comps
            lower_theor = theor_comps
        else
            upper_mult = i
            upper_comp = est_comps
            upper_theor = theor_comps
            break
        end
    end
    if upper_mult == lower_bounds
        (est_comps, theor_comps) = estimate_comparisons(genomes, lower_bounds)
    elseif lower_mult == upper_bounds
        return upper_bounds
    else
        lower_diff = comparisons - lower_comp
        upper_diff = upper_comp - comparisons
        if ldiff <= udiff
            return lower_mult
        else
            return upper_mult
        end
    end
end

function estimate_comparisons(genomes::Integer, mult::Integer)
    # Estimate the number of comparisons necessary for good bootstrapping
    total_comp = 0
    theor_comp = 0
    for i = range(2,genomes+1)
        theor = get_theoretical_comps(genomes, i)
        real = mult * genomes
        theor_comp += theor
        total_comp += min(theor, real)
    end
    return (total_comp, theor_comp)
end

function get_theoretical_comps(genomes::Integer, i::Integer)
    # Get number of combinations for the given size pangenome
    factorial(genomes) / (factorial(genomes - i) * factorial(i-1))
end

function create_reference_string(reference_set)
    # Create a string from a random set of genomes
    seen_vec = [ reference_set[i] for i = 1:length(reference_set) ]
    sort!(seen_vec)
    join(seen_vec, ":")
end

function categorize_shared_gene(genes_by_category, comp_genome::String, gene::String)
    # Increment shared count for this gene if we have a hit
    get!(genes_by_category[comp_genome]["shared"], gene, 0)
    genes_by_category[comp_genome]["shared"][gene] += 1
    return genes_by_category
end

function categorize_core_gene(genes_by_category, comp_genome::String, gene::String)
    # Determine if the given gene is present in all other genomes up to this point.
    genes_by_category[comp_genome]["core"][gene] = 1
    return genes_by_category
end

function categorize_new_gene(genes_by_category, comp_genome::String, gene::String)
    # Determine if the given gene is present in no other genomes up to this point.
    genes_by_category[comp_genome]["new"][gene] = 1
    return genes_by_category
end

function write_single_genomes(hp::HitsProfile, fh)
    # Write the number of genes for the individual genomes
    genome_index = Dict{String, Integer}(hp.genome_names[i] => i for i in 1:hp.num_genomes)

    # For each genome, get number of genes and write to file (also write assigned index)
    for genome in hp.genome_names
        num_genes = length(keys(hp.profile[genome]))
        index = genome_index[genome]
        write(fh, "1\t0\t0\t$num_genes\t$index\t$genome\n")
    end
end

function write_results(genes_by_category, comp_genome::String, size::Integer, hp::HitsProfile, fh)
    index = findfirst(hp.genome_names,comp_genome)
    p_size = size + 1 # Pangenome size = Ref genomes + seed genome

    # Write core/shared/new results to file
    shared_count = length(keys(genes_by_category[comp_genome]["shared"]))
    new_count = length(keys(genes_by_category[comp_genome]["new"]))
    core_count = length(keys(genes_by_category[comp_genome]["core"]))

    write(fh, "$p_size\t$core_count\t$shared_count\t$new_count\t$index\t$comp_genome\n")
end

function do_analysis_with_sampling(args, hp::HitsProfile, fh)
    multiplex = determine_sampling_rate(hp, args)

    # Iterate through each possible pangenome size
    for i = 1:(hp.num_genomes-1)
        tic()
        print("Running analysis with $i genomes...")
        sampling_counter = Dict{String, Integer}()

        fac_numer = factorial(big(hp.num_genomes-1))    # Must convert to BigInt first
        fac_denom1 = factorial(big((i+1)-1))            # ... otherwise OverflowError
        fac_denom2 = factorial(big(hp.num_genomes-(i+1)))
        true_max = fac_numer / (fac_denom1 * fac_denom2)

        # Iterate through the all genomes in set
        for j = 1:hp.num_genomes
            seen_strings = Dict{String, Bool}()
            seed_genome = hp.genome_names[j]
            # Create group of reference genomes minus the seed genome
            ref_genomes = copy(hp.genome_names) # copying the variable creates a pointer
            splice!(ref_genomes, j)
            point_count = 0
            while point_count < min(multiplex, true_max)
                # Create of string of N-reference genomes for an N+1 sized pangenome
                reference_set = rand(ref_genomes, i)
                sort_string = create_reference_string(reference_set)

                # Proceed if this is a new string
                if haskey(seen_strings, sort_string) == false
                    get!(sampling_counter, seed_genome, 1)
                    genes_by_category = Dict()
                    genes_by_category[seed_genome] = Dict("shared" => Dict(), "new" => Dict(), "core" => Dict())

                    # Iterate through all the genes in the seed genome
                    for gene in keys(hp.profile[seed_genome])
                        for ref_g in reference_set
                            # If reference genome is a key in the HitsProfile, add to "shared" genes count
                            if haskey(hp.profile[seed_genome][gene], ref_g)
                                genes_by_category = categorize_shared_gene(genes_by_category, seed_genome, gene)
                            end
                        end
                        # Determine if gene is shared by all or none of the referece genomes
                        count = get(genes_by_category[seed_genome]["shared"], gene, 0)
                        if count == length(reference_set)
                            genes_by_category = categorize_core_gene(genes_by_category, seed_genome, gene)
                        elseif count == 0
                            genes_by_category = categorize_new_gene(genes_by_category, seed_genome, gene)
                        end
                    end
                    write_results(genes_by_category, seed_genome, i, hp, fh)
                    sampling_counter[seed_genome] += 1
                    seen_strings[sort_string] = true
                    point_count += 1
                end
            end
        end

        elapsed = toq()
        println("time elapsed: $elapsed secs")
    end

end

function do_analysis_without_sampling()
    println("Doing analysis without sampling is not implemented at this time")
end

function parse_commandline()
    s = ArgParseSettings(description = "Reads a pangenome profile matrix and generates total genes in the pangenome",
    prog = "pangenome_make_pangenome.jl",
	version = "1.0",
	add_version = true,
	add_help = true)

    # The macro to add a table of arguments and options to the ArgParseSettings object
    @add_arg_table s begin
        "--profile", "-p"
            help = "Path to profile matrix"
            metavar = "/path/to/pangenome/profile.txt"
            required = true
        "--output_dir", "-o"
            help = "Directory to write the output"
            metavar = "/path/to/pangenome/dir/"
            required = true
        "--comparisons", "-c"
            help = "The number of comparisons to make for any 1 value of N (sampling)"
            metavar = "100000"
			default = 0
			range_tester = (x->x>=0)
            arg_type = Int
        "--multiplicity", "-m"
            help = "Another option for sampling based on a multiplicity factor ((sum(m*n) for n=[2..n]) number of comparisons)"
            metavar = "20"
			default = 0
			range_tester = (x->x>=0)
            arg_type = Int
    # Will not add log_file or debug options for now
    end

    # Converts the ArgParseSettings object into key/value pairs
    return parse_args(s)
end

function main()
    args = parse_commandline()
    println("Parsed args:")
    for (arg,val) in args
        println("  $arg  =>  $val")
    end

    hp = HitsProfile()
    read_in_profile(hp, args["profile"])

    output_dir = args["output_dir"]
    output_file = "$output_dir/core_shared_new.tsv"
    out_fh = open(output_file, "w")

    ## our output table will take the form:
    ## genome_count  core    shared    new  root_genome
    write_single_genomes(hp, out_fh)

    if args["comparisons"] == 0 && args["multiplicity"] == 0
        do_analysis_without_sampling()
    else
        do_analysis_with_sampling(args, hp, out_fh)
    end

    close(out_fh)
end

main()
quit()
